#150 Provide Client API V2 and Client implementation

CamilleLetavernier opened the pull request:
Note: the PR is based on the standalone client PR #69 

This is the client changes to support the V2 ModelServer API. The changes are described on the corresponding Model Server pull request: eclipse-emfcloud/emfcloud-modelserver/pull/151

Contributed on behalf of STMicroelectronics.

CamilleLetavernier commented:
Current state:

- Some commands are untested. Since the server currently only supports 'replace' operations, this is the only tested operation.
- Json Patch notifications are available in the API, but they are not automatically handled by the Client implementation. I still need to apply the patch on the current client-side Json Model, to get the incrementally-updated model (We'll rely on the `fast-json-patch` library for this)

CamilleLetavernier commented:
> Some commands are untested. Since the server currently only supports 'replace' operations, this is the only tested operation.

Added implementation & tests for Add (Create) and Remove operations. Move and Test are still unsupported.

CamilleLetavernier commented:
The pull request has been rebased on `master`, to properly include the changes from #65

When invoking edit() with a Patch or a Command, the resulting Json Patch diff can now be immediately applied on the model. Since the Client API itself doesn't know the current model state (and thus can't directly patch it), it returns a function that can be applied on the model instead. It's the responsibility of the client of this API to decide how to apply the patch (or whether it should be applied at all).

Other options are still possible: querying the full model from the model server, or waiting for an update notification.

CamilleLetavernier commented:
Added patch functions for:

- Undo/Redo results
- Model update notifications

In addition to the patch function, the actual Json Patch object has also been added to the result or notification, so clients can inspect it, or maybe manipulate it with a different json-patch library. 

Also added a "copy" boolean parameter to the patch function, so clients can decide whether to directly apply the patch on the current model (in-place modification), or on a copy of the model.

ndoschek commented on packages/modelserver-client/src/model-server-client-api-v2.ts:
```diff
@@ -0,0 +1,133 @@
+    /**
+     * Retrieves all available models of the current workspace.
+     */
+    getAll(): Promise<Model[]>;
```
It would be great if the V2 API would be documented as well from the beginning. I assume most methods can reuse the documentation from V1.

ndoschek commented on packages/modelserver-client/src/utils/patch-utils.ts:
```diff
@@ -0,0 +1,188 @@
+import { ModelServerObjectV2, ModelServerReferenceDescriptionV2 } from '../model/base-model';
+import { AnyObject, TypeGuard } from './type-util';
+
+// Utility methods to create Json Patches
```
```suggestion
```

ndoschek reviewed (commented):
Thanks Camille, changes look great to me!

Regarding the tests, the launch config `Run all tests` only executes two test files in my case - `tests for Diagnostic` and `tests for type-util`. The `test` script in `packages/modelserver-client` executes all avaiable specs. Can you observe the same behavior?

CamilleLetavernier commented:
> Regarding the tests, the launch config `Run all tests` only executes two test files in my case - `tests for Diagnostic` and `tests for type-util`. The `test` script in `packages/modelserver-client` executes all avaiable specs. Can you observe the same behavior?

Indeed, there was a configuration issue: The launc config was looking for files under `src/**/*.spec.ts`, which ignored test files located directly under `src/`

I've updated the launch configs. I've also added a separate config for "All unit tests" (in addition to "All tests"), that ignores the new integration tests (So we can still run some tests without having the model server running).

ndoschek reviewed (changes requested):
Thanks Camille, the test launch configs work great now!

There is one thing I noticed, the custom menu which sends custom requests to the custom counter endpoint do not work anymore. I assume the request path has to be updated to `/v2//` as well. Could you please have another look?

Besides that, the PR could be merged imo, and I suggest to add Martin as a final reviewer then :)

ndoschek reviewed (approved):
Thanks Camille, LGTM! :+1: 
Will add @martin-fleck-at as a final reviewer before merging it.

ndoschek commented:
Part of eclipse-emfcloud/emfcloud-modelserver/issues/155

martin-fleck-at commented on packages/modelserver-client/src/model-server-client-v2.ts:
```diff
@@ -0,0 +1,338 @@
+/**
+ * A client to interact with a model server.
+ */
+export class ModelServerClientV2 implements ModelServerClientApiV2 {
```
I saw on the server side that there is support for transactions. Should that be reflected in the client as well?

CamilleLetavernier replied:
Copying the reply from the other PR:

> It's used by a new project: https://github.com/eclipse-emfcloud/modelserver-node
> 
> We decided to exclude it from the Javascript/Java client API at the moment; it's considered an implementation detail, to > be used through the modelserver-node APIs (At the moment. But maybe it could be properly opened in the future).

martin-fleck-at replied:
I understand, so it is more internal API that is not yet supposed to be "public". In that case I am fine with not exposing it for now. However, usually I am not a fan of hiding functionality that is obviously there so it might make more sense to explicitly mark it internal through comments or a second interface. But this should not prevent this PR from being merged.

CamilleLetavernier commented on packages/modelserver-client/src/utils/patch-utils.ts:
```diff
@@ -0,0 +1,209 @@
+
+    /** Type guard testing whether an operation is a remove operation. */
+    export function isRemove(op: Operation): op is RemoveOperation {
+        return op?.op === 'replace';
```
Good catch! :+1:

CamilleLetavernier commented on packages/modelserver-client/src/model-server-client-api-v2.ts:
```diff
@@ -0,0 +1,133 @@
+export type JsonFormat = 'json' | 'json-v2';
+
+/** Message formats supported by the V2 client API. */
+export type Format = 'xml' | JsonFormat;
```
Indeed, it should be xmi; and I agree with the additional `string` type (Although it kinda defeats the purpose of a custom Format type in the first place).

@cdamus WDYT?

cdamus replied:
Yes, that a good point about custom formats. Extenders could even have a need for custom JSON formats. I think in that light it's probably best just to replace the `JsonFormat` type with `string`.

CamilleLetavernier replied:
Sounds good. I think the 'xml' format was coming from `model-server-client.spec.ts`, which tests arbitrary REST requests (which are not actually interpreted by anyone). So using a random format made sense; but xml is not actually supported by the server.

When introducing constants for the known formats (xmi, json and json-v2), I realized that our V2 Client still uses 'json' as the default format. Should it be 'json-v2' instead? (The server has already been updated to use 'json-v2' as the new default).

martin-fleck-at replied:
I think we may even have left the constants defined as is and just added the 'string' type but simply using string and other constants is absolutely fine as well. I also think it was a good idea to update to json-v2 for the v2 client to match the server, thank you Camille!

CamilleLetavernier commented on packages/modelserver-client/src/model-server-client-api-v2.ts:
```diff
@@ -0,0 +1,133 @@
+/**
+ * Basic client API to interact with a model server that conforms to the Modelserver API Version 2.
+ */
+export interface ModelServerClientApiV2 {
```
I'm not sure. Most of the interfaces are not duplicated (at the moment), as V1 and V2 are still 95% similar (The command execution and notifications parts are the only big differences).

Splitting namespaces might introduce additional inconsistencies or complexity, that we don't necessarily need (or want). For example, we'd have a "Common" (or empty) namespace for APIs shared by V1 and V2, and then "V1" and "V2" specific namespaces. What if we need to specialize one more API for V2? Would we move it from "common" to "v1" and duplicate it in "v2"? Would that break existing V1/V2 clients?

And duplicating everything "just in case" is something we want to avoid (See eclipse-emfcloud/emfcloud-modelserver/pull/169)

martin-fleck-at replied:
I see, I think for now having the classes as they are is fine for now. I just was interested to see what the reasoning behind it was. Of course, in the long run we may need to find a better approach if we want to support multiple versions but that is yet to be seen. Since both v1 and v2 still work as intended, this is fine for me.

CamilleLetavernier commented on packages/modelserver-client/src/utils/patch-utils.ts:
```diff
@@ -0,0 +1,209 @@
+    };
+}
+
+export function getObjectPath(modeluri: string, object: ModelServerObjectV2 | ModelServerReferenceDescriptionV2): string {
```
Hmm. Why is this even exported? It's an implementation detail of the higher-level functions defined above. I think it should be local only (But it can still be documented :) especially since we now support both EMF-like paths and standard Json Pointer paths)

Now I need to remember why I exported them in the first place... I'm pretty sure they weren't meant to be.

martin-fleck-at replied:
Removing export AND provide documentation, what a treat, thank you ;-)

CamilleLetavernier commented on packages/modelserver-client/src/utils/type-util.ts:
```diff
@@ -145,3 +146,172 @@ export function asModelArray(object: unknown): Model[] {
+}
+
+function isNonEmptyObjectArray(value: any): value is AnyObject[] {
+    // TODO: More correct would be value.every(e => typeof e === 'object' && !Array.isArray(value[0]))
```
@cdamus ?

cdamus replied:
I emulated examples found elsewhere in the EMF.Cloud universe, but then added my own commentary. I'm happy either to remove the comment or to implement its suggestion, except that the suggestion I think is actually wrong: it doesn't actually reject an empty array because `every` is always `true` for empty arrays.

martin-fleck-at replied:
So in that case, I would suggest to simply remove the comment for now unless we actually run into problems with it.
@CamilleLetavernier , @cdamus Could either of you adapt the PR or accept the suggestion?
```suggestion
```

CamilleLetavernier commented on packages/modelserver-client/src/subscription-listener.ts:
```diff
@@ -157,3 +160,53 @@ export class NotificationSubscriptionListener implements SubscriptionListener {
+ * A {@link ModelServerNotificationListener} for V2 Client API. Uses JsonPatch for incremental
+ * update notifications, instead of CommandExecutionResults.
+ */
+export interface ModelServerNotificationListenerV2 extends ModelServerNotificationListener {
```
Thanks, I wasn't aware of this class. I'll add a similar one for V2 :+1:

martin-fleck-at replied:
Wonderful, thank you!

CamilleLetavernier commented:
I pushed two new commits which should address most comments:

- Use constants for Json Patch 'op' types
- Fix the 'isRemove' function to use the correct 'op' type
- Improve jsdoc comments
- Fix incorrect format 'xml'; should be 'xmi'
- Change the "Format" type to allow any string. Use constants for known supported formats.
- Add support for V2 Incremental Update Notifications in ModelServerSubscriptionClient

Some things to consider:

- I changed the "default format" for the V2 Client from 'json' to 'json-v2', since 'json' is the "V1" format. This should be more consistent (The ModelServer also uses V2 as the default format)
- I've added a ModelServerSubscriptionClientV2 class. Since both V1 and V2 Subscriptions use the same message type (incrementalUpdate), we can't distinguish them at runtime. Clients have to know whether they want "V1" or "V2" notifications (APIs for Notifications are actually separated, but the raw json messages uses the same type in both cases). Since the "V1" interface is the same, but the behavior is different, I'm not sure how to best reflect this in the frontend module. I've added a new frontend module for V2 (which should override/complement the V1 Module), but I don't know how to properly handle this. You can't use the V1 and the V2 implementation at the same time, and using the V2 Implementation will break V1 notifications; so clients need to make a consistent choice here (Bind the V2 implementation if they use the V2 API, or bind the V1 implementation if they use the V1 API).

martin-fleck-at reviewed (approved):
@CamilleLetavernier Thank you very much for the updates!

Having json-v2 as default format of v2 seems reasonable to me.

I think it is okay that the client has to make a decision between V1 and V2 at the moment since I do not believe people will be using them in parallel anyway.

I had one small suggestion left (to remove a comment) which would be great if you could merge into the PR. Other than that, I think this can be merged! Fantastic work @CamilleLetavernier @cdamus! I am having another look at the server-side changes.

martin-fleck-at commented:
The build does not run through at the moment. Locally, everything works fine. I also triggered a master build and saw the same problem there:
```
error /home/jenkins/agent/workspace/mfcloud-modelserver-theia_master/node_modules/vscode-ripgrep: Command failed.
Exit code: 1
Command: node ./lib/postinstall.js
Arguments: 
Directory: /home/jenkins/agent/workspace/mfcloud-modelserver-theia_master/node_modules/vscode-ripgrep
Output:
Finding release for v13.0.0-4
GET https://api.github.com/repos/microsoft/ripgrep-prebuilt/releases/tags/v13.0.0-4
Downloading from https://api.github.com/repos/microsoft/ripgrep-prebuilt/releases/assets/47921394
Downloading to /tmp/vscode-ripgrep-cache-1.13.2/ripgrep-v13.0.0-4-x86_64-unknown-linux-musl.tar.gz
Download options: {"headers":{"user-agent":"vscode-ripgrep","accept":"application/octet-stream"}}
statusCode: 302
Following redirect to: https://objects.githubusercontent.com/github-production-release-asset-2e65be/194786020/3c56f800-b252-43d7-bf23-e91be99d74dc?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20220302%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20220302T124003Z&X-Amz-Expires=300&X-Amz-Signature=ecd9bc8b865af7530b31e727b552146f4a4a045ac8519815dfa250e4618f8197&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=194786020&response-content-disposition=attachment%3B%20filename%3Dripgrep-v13.0.0-4-x86_64-unknown-linux-musl.tar.gz&response-content-type=application%2Foctet-stream
Download options: {"headers":{"user-agent":"vscode-ripgrep","accept":"application/octet-stream"}}
Deleting invalid download cache
Downloading ripgrep failed: Error: getaddrinfo EAI_AGAIN objects.githubusercontent.com
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:66:26)
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

So we probably can still merge this PR if we do not manage to resolve this issue soon.

CamilleLetavernier commented:
> I had one small suggestion left (to remove a comment) which would be great if you could merge into the PR.

Done :+1:

I've squashed the commits and rebased on master
